### ISSUE TITLE
NPE in REST API when updating thing with channels that have no ChannelTypeUID

### ISSUE THREAD

**Issue by mhilbush** [openhab/openhab-core#1457]
Prerequisite:
You have a thing that has at least one channel that has no `ChannelTypeUID`

When editing the `location` of a managed thing using Paper UI, you get a 500 Internal Server Error. Looking at the response from openHAB, it appears something in OHC is throwing a NPE. Unfortunately, I don't see the full stack trace in the log file.
```
{"error":{"http-code":500,"exception":{"class":"java.lang.NullPointerException"}}}
```

I notice that when performing the update from Paper UI, Paper UI sends the entire thing object in the thing update request (including the dynamically created channels).

Performing the same operation using the rest API works if you send just this payload.
```
{ "location": "somewhere" }
```

It must be happening somewhere from this line down, since the errors above this line are not returned.
https://github.com/openhab/openhab-core/blob/d4b0ecb122b2c7414572f61475780b712bf4cf22/bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/thing/ThingResource.java#L375

I don't know whether the new 3.0 admin UI will work the same way as Paper UI (i.e. does it send the entire thing object, or just what is changed).

@ghys how will this work in the new admin UI? Should I try to track down the root cause of the above issue, or is it moot for 3.0?

For completeness, here's the payload sent by Paper UI.
```
{"statusInfo":{"status":"ONLINE","statusDetail":"NONE"},"editable":true,"label":"Ecobee Sensor Family Room","bridgeUID":"ecobee:thermostat:server:xxxxx","configuration":{"sensorId":"rs:100"},"properties":{},"UID":"ecobee:sensor:server:xxxxx:rs-100","thingTypeUID":"ecobee:sensor","channels":[{"linkedItems":[],"uid":"ecobee:sensor:server:xxxxx:rs-100:id","id":"id","channelTypeUID":"ecobee:sensorId","itemType":"String","kind":"STATE","label":"Sensor Id","defaultTags":[],"properties":{},"configuration":{}},{"linkedItems":[],"uid":"ecobee:sensor:server:xxxxx:rs-100:name","id":"name","channelTypeUID":"ecobee:sensorName","itemType":"String","kind":"STATE","label":"Sensor Name","defaultTags":[],"properties":{},"configuration":{}},{"linkedItems":[],"uid":"ecobee:sensor:server:xxxxx:rs-100:type","id":"type","channelTypeUID":"ecobee:sensorType","itemType":"String","kind":"STATE","label":"Sensor Type","defaultTags":[],"properties":{},"configuration":{}},{"linkedItems":[],"uid":"ecobee:sensor:server:xxxxx:rs-100:code","id":"code","channelTypeUID":"ecobee:sensorCode","itemType":"String","kind":"STATE","label":"Sensor Code","defaultTags":[],"properties":{},"configuration":{}},{"linkedItems":[],"uid":"ecobee:sensor:server:xxxxx:rs-100:inUse","id":"inUse","channelTypeUID":"ecobee:sensorInUse","itemType":"Switch","kind":"STATE","label":"Sensor In Use","defaultTags":[],"properties":{},"configuration":{}},{"linkedItems":[],"uid":"ecobee:sensor:server:xxxxx:rs-100:temperature","id":"temperature","itemType":"Number:Temperature","kind":"STATE","label":"Sensor Temperature","defaultTags":[],"properties":{},"configuration":{}},{"linkedItems":[],"uid":"ecobee:sensor:server:xxxxx:rs-100:occupancy","id":"occupancy","itemType":"Switch","kind":"STATE","label":"Sensor Occupancy","defaultTags":[],"properties":{},"configuration":{}}],"location":"somewhere"}
```

**Comment by cdjackson** [comment]
> how will this work in the new admin UI? Should I try to track down the root cause of the above issue, or is it moot for 3.0?

I would argue that even if the UI in OH3 is doing something different, the core should not NPE as other UIs may be used (or there are plenty of people just using the REST API directly). So, I think it would be best to track it down if possible...

**Comment by mhilbush** [comment]
Can't disagree with that, Chris. I wish I could get a stack trace, as that would make it dramatically easier to pinpoint the problem...

**Comment by ghys** [comment]
Agreed, and I don't think the new UI will be any different anyways... it'll
post the entire thing object because it's convenient to have an atomic
operation saving everything on the page

**Comment by cdjackson** [comment]
I know this is a big problem throughout the REST interfaces - most exceptions are simply swallowed. In some places people have added code to handle this better, so that might be the best option for starters - if you can find out where the exception is caught, then log it - possible in the REST endpoint itself?

> I don't think the new UI will be any different anyways... it'll post the entire thing object because it's convenient 

That's a shame :( We'll still have the same limitations then.  (but that's a different discussion).

**Comment by mhilbush** [comment]
Calling the RESP API with the whole object is a little more illuminating. Not a stack trace, but probably good enough.
```
{
  "error": {
    "message": "Channel UID 'ecobee:sensor:server:xxxxx:rs-100:id' does not match thing UID 'ecobee:sensor:server:313880787137:rs-100'",
    "http-code": 500,
    "exception": {
      "class": "java.lang.IllegalArgumentException",
      "message": "Channel UID 'ecobee:sensor:server:xxxxx:rs-100:id' does not match thing UID 'ecobee:sensor:server:313880787137:rs-100'",
      "localized-message": "Channel UID 'ecobee:sensor:server:xxxxx:rs-100:id' does not match thing UID 'ecobee:sensor:server:313880787137:rs-100'"
    }
  }
}
```

**Comment by ghys** [comment]
I don't think the new UI will be any different anyways... it'll post the
> entire thing object because it's convenient
>
> That's a shame :( We'll still have the same limitations then. (but that's
> a different discussion).
>
That's quite an abridged quote.

>

**Comment by mhilbush** [comment]
Probably coming from here?

https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/builder/ThingBuilder.java#L152

**Comment by cdjackson** [comment]
> That's quite an abridged quote.

Sorry - it wasn't meant to be. I simply meant it's a shame that the data will always all be sent since it causes problems. Convenience is irrelevant - it's just how the cut and paste went and wasn't meant to imply anything in itself.

**Comment by mhilbush** [comment]
One other data point. Using the rest API, if I send the entire thing object minus the two dynamically created channels, the update is successful. So, I think that at least narrows it down to where it's iterating through the channels (which I guess should not be surprising).

**Comment by mhilbush** [comment]
As the channels have no configuration, it's probably not on this call.

https://github.com/openhab/openhab-core/blob/d4b0ecb122b2c7414572f61475780b712bf4cf22/bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/thing/ThingResource.java#L377

**Comment by mhilbush** [comment]
I think my previous statement was wrong.

I'll bet the NPE is happening here when it tries to use `channelbean.channelTypeUID`.
https://github.com/openhab/openhab-core/blob/d4b0ecb122b2c7414572f61475780b712bf4cf22/bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/thing/ThingResource.java#L892

One thing I see is that the dynamically created channels don't have a ChannelTypeUID. So that would seem to cause an NPE in the above line.

Here's the channel definition from the JSONDB. As you can see, there's no ChannelTypeUID.
```
        {
          "acceptedItemType": "Number:Temperature",
          "kind": "STATE",
          "uid": {
            "segments": [
              "ecobee",
              "sensor",
              "server",
              "313880787137",
              "rs-100",
              "temperature"
            ]
          },
          "label": "Sensor Temperature",
          "configuration": {
            "properties": {}
          },
          "properties": {},
          "defaultTags": []
        },
```

@cweitkamp @kaikreuzer I'd benefit from your guidance here.

My binding creates some channels dynamically here. These channels don't have a ChannelTypeUID.
https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/handler/EcobeeSensorThingHandler.java#L122

However, during an update of the thing, the OHC code referenced above is trying to access the ChannelTypeUID of the channels that were passed in on the REST call to update the thing.

Is it ok to create a channel without a ChannelTypeUID? OHC certainly allows it. However, it appears to cause problems. 

If it's not ok, should it be enforced somewhere?

The easy fix for my binding is to call `withType` when building the channels, but that doesn't really solve the overall problem.

**Comment by cdjackson** [comment]
> One thing I see is that the dynamically created channels don't have a ChannelTypeUID. So that would seem to cause an NPE in the above line.

I just had a look at the ZigBee binding, and I would have expected there to be a `ChannelTypeUID`. In the `ChannelBuilder`, there is a `withType` method which sets the `ChannelTypeUID`. Maybe there's a bug and this isn't used (although from a quick look, it does seem to be) or the binding in question isn't setting the channel type when it creates the channels (which would be a bug in the binding, and there should be a check in the core).

From the ZigBee binding -:

```
        return ChannelBuilder
                .create(createChannelUID(thingUID, endpoint, ZigBeeBindingConstants.CHANNEL_NAME_POWER_BATTERYVOLTAGE),
                        ZigBeeBindingConstants.ITEM_TYPE_NUMBER)
                .withType(ZigBeeBindingConstants.CHANNEL_POWER_BATTERYVOLTAGE)
                .withLabel(ZigBeeBindingConstants.CHANNEL_LABEL_POWER_BATTERYVOLTAGE)
                .withProperties(createProperties(endpoint)).build();
```

and from ChannelBuilder` -:

```
    /**
     * Appends the channel type to the channel to build
     *
     * @param channelTypeUID channel type UID
     * @return channel builder
     */
    public ChannelBuilder withType(@Nullable ChannelTypeUID channelTypeUID) {
        this.channelTypeUID = channelTypeUID;
        return this;
    }
```

**Comment by mhilbush** [comment]
Yeah, I saw that. I'm not calling it in my ChannelBuilder, and there's no enforcement that it's missing. I'm adding a call to `withType` now to see if it fixes the problem.

**Comment by mhilbush** [comment]
Yep, that eliminated the NPE. What's odd is that everything works fine except for this issue.

I do think that something should change in core. Either enforce that the channel must have a Channel Type UID when building a channel, or be more resilient in the REST API. I just don't know enough to say whether it should be the former or the latter.

WDYT?

**Comment by cdjackson** [comment]
I personally think that the channel should have a type, and this should be enforced. Even in a dynamic system, the binding knows at compile time what channels it will support (I assume?). I would fear there are otherwise other places where the system makes the assumption that the type exists...

Also, if a binding defines a channel in the thing type via XML that doesn't exist it will throw an exception. I appreciate that's a slightly different situation as it needs to find the channel type in the registry, but there should be consistency (IMHO)...

**Comment by cdjackson** [comment]
> Even in a dynamic system, the binding knows at compile time what channels it will support (I assume?).

And if it doesn't, then I would think it should implement its own `ChannelTypeProvider` so that the type is still registered.

**Comment by mhilbush** [comment]
> the binding knows at compile time what channels it will support

Yes, in this case the channel types are well known at compile time.

**Comment by mhilbush** [comment]
Interestingly, in `ChannelBuilder`, `channelTypeUID` is defined as `@Nullable`.
https://github.com/openhab/openhab-core/blob/61e17ce39e83daa7f1d8c41512dead8d95d6e5d9/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/builder/ChannelBuilder.java#L47

And, `withType` allows a null argument.
https://github.com/openhab/openhab-core/blob/61e17ce39e83daa7f1d8c41512dead8d95d6e5d9/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/builder/ChannelBuilder.java#L96

Seems it was a conscious decision to allow no `ChannelTypeUID` when building a channel?

**Comment by mhilbush** [comment]
In any event, I pushed a fix for it in my binding. 😉

**Comment by mhilbush** [comment]
I updated the title and initial post to be more reflective of the root cause of the issue.

**Comment by mhilbush** [comment]
Now that I look at this further, I don't think it can be in `normalizeChannels`. There's a test for null right above the statement that I thought was causing the problem. 

What I do know is that my change to add a `ChannelTypeUID` to the dynamically created channels resolves the NPE issue.

**Comment by mhilbush** [comment]
Relevant portion of stack trace...

```
java.lang.NullPointerException: null
	at org.eclipse.smarthome.core.common.AbstractUID.splitToSegments(AbstractUID.java:113) ~[bundleFile:?]
	at org.eclipse.smarthome.core.common.AbstractUID.<init>(AbstractUID.java:49) ~[bundleFile:?]
	at org.eclipse.smarthome.core.thing.UID.<init>(UID.java:48) ~[bundleFile:?]
	at org.eclipse.smarthome.core.thing.type.ChannelTypeUID.<init>(ChannelTypeUID.java:40) ~[bundleFile:?]
	at org.eclipse.smarthome.core.thing.dto.ChannelDTOMapper.map(ChannelDTOMapper.java:52) ~[bundleFile:?]
	at org.eclipse.smarthome.core.thing.util.ThingHelper.merge(ThingHelper.java:229) ~[bundleFile:?]
	at org.eclipse.smarthome.io.rest.core.internal.thing.ThingResource.update(ThingResource.java:385) [bundleFile:?]
```

**Comment by mhilbush** [comment]
Ack! This is already fixed in 3.0!

https://github.com/openhab/openhab-core/pull/1346
